src/internal/ppsutil/util.go

```diff
@@ -381,6 +381,57 @@ func UpdateJobState(pipelines col.ReadWriteCollection, jobs col.ReadWriteCollect
 	return jobs.Put(jobPtr.Job.ID, jobPtr)
 }
 
+func FinishJob(pachClient *client.APIClient, jobInfo *pps.JobInfo, state pps.JobState, reason string) error {
+	jobInfo.State = state
+	jobInfo.Reason = reason
+	var empty bool
+	if state == pps.JobState_JOB_FAILURE || state == pps.JobState_JOB_KILLED {
+		empty = true
+	}
+	_, err := pachClient.RunBatchInTransaction(func(builder *client.TransactionBuilder) error {
+		if _, err := builder.PfsAPIClient.FinishCommit(pachClient.Ctx(), &pfs.FinishCommitRequest{
+			Commit: jobInfo.OutputCommit,
+			Empty:  empty,
+		}); err != nil {
+			return err
+		}
+		if _, err := builder.PfsAPIClient.FinishCommit(pachClient.Ctx(), &pfs.FinishCommitRequest{
+			Commit: jobInfo.StatsCommit,
+			Empty:  empty,
+		}); err != nil {
+			return err
+		}
+		return WriteJobInfo(&builder.APIClient, jobInfo)
+	})
+	return err
```

Comment:
Yeah, I removed that because I actually think we should since that would be a critical error (particularly if the recovery code is not run). We should strongly enforce that all of the state is consistent, and make it very clear when that is not the case. It should not be possible to finish the output commit without finishing the corresponding stats commit. We may need to block the finishing of pipeline output commits, and force the use of stop job.

Reply:
This doesn't do anything to ensure that external code can't violate the invariants - output commits may still be closed at will because PFS doesn't (and probably shouldn't) deal with PPS logic.  What are the recovery steps a user has to take if they manage to break this code?

Reply:
Yeah, we need to address that (I already have a related TODO for starting commits that I updated). I think we should push for preventing that rather than recovering from it though. Not sure what the best approach would be, but I think it would likely be scoping the starting / finishing of certain commits (in a repo or branch) in such a way that a user / outside service would not be able to start / finish commits where they shouldn't.